integrations/pgvector/src/haystack_integrations/document_stores/pgvector/document_store.py

```diff
@@ -0,0 +1,377 @@
+# SPDX-FileCopyrightText: 2023-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+import logging
+from typing import Any, Dict, List, Literal, Optional
+
+from haystack import default_to_dict
+from haystack.dataclasses.document import ByteStream, Document
+from haystack.document_stores.errors import DocumentStoreError, DuplicateDocumentError
+from haystack.document_stores.types import DuplicatePolicy
+from psycopg import Error, IntegrityError, connect
+from psycopg.abc import Query
+from psycopg.cursor import Cursor
+from psycopg.rows import dict_row
+from psycopg.sql import SQL, Identifier
+from psycopg.sql import Literal as SQLLiteral
+from psycopg.types.json import Json
+
+from pgvector.psycopg import register_vector
+
+logger = logging.getLogger(__name__)
+
+TABLE_DEFINITION = [
+    ("id", "VARCHAR(128)", "PRIMARY KEY"),
+    ("embedding", "VECTOR({embedding_dimension})", ""),
+    ("content", "TEXT", ""),
+    ("dataframe", "JSON", ""),
+    ("blob_data", "BYTEA", ""),
+    ("blob_meta", "JSON", ""),
+    ("blob_mime_type", "VARCHAR(255)", ""),
+    ("meta", "JSON", ""),
+]
+
+SIMILARITY_FUNCTION_TO_POSTGRESQL_OPS = {
+    "cosine_distance": "vector_cosine_ops",
+    "max_inner_product": "vector_ip_ops",
+    "l2_distance": "vector_l2_ops",
+}
+
+HNSW_INDEX_CREATION_VALID_KWARGS = ["m", "ef_construction"]
+
+HNSW_INDEX_NAME = "haystack_hnsw_index"
+
+
+class PgvectorDocumentStore:
+    def __init__(
+        self,
+        *,
+        connection_string: str,
+        table_name: str = "haystack_documents",
+        embedding_dimension: int = 768,
+        embedding_similarity_function: Literal[
```

Comment:
I'm not sure about this... We're introducing new names like `cosine_distance` that don't exist in pgvector, so they need to be documented, won't be part of any docs outside Haystack and won't be familiar to pgvector users coming to Haystack. I think changing this parameter to `postgresql_ops` or `pgvector_ops`  and having the literal set to `"vector_cosine_ops", "vector_ip_ops", "vector_l2_ops",`  would be more explicit.

Reply:
- these names should be familiar for pgvector users: see https://github.com/pgvector/pgvector?tab=readme-ov-file#vector-functions

- regarding `embedding_similarity_function`, it is more of a Haystack concept... Can we call it `vector_distance`?

Reply:
What about `vector_function` with reference to https://github.com/pgvector/pgvector?tab=readme-ov-file#vector-functions ?

Reply:
OK. Let's do that